cpp/ql/src/jsf/4.10 Classes/AV Rule 82.ql

```diff
@@ -72,7 +72,10 @@ predicate assignOperatorWithWrongType(Operator op, string msg) {
 predicate assignOperatorWithWrongResult(Operator op, string msg) {
   op.hasName("operator=")
   and not returnsDereferenceThis(op)
-  and exists(op.getBlock())
+  // If a function does not have a reachable `ReturnStmt` then either its body
+  // was not in the snapshot or it was established by the extractor or the CFG
+  // pruning that the function never returns.
+  and exists(ReturnStmt ret | ret.getEnclosingFunction() = op and reachable(ret))
```

Comment:
Would it be better to put the `reachable` check in `returnsDereferenceThis`?

As it stands, I suspect it will do the wrong thing if there are two `ReturnStmt`s, one which returns `*this` and the other a compiler generated return that's unreachable?

Reply:
You're right, the check should be in the predicates where we quantify over `ReturnStmt`. I can't see a problem with the example you mention, but I added tests to show three more FPs and then a commit to fix them. Performance looks good: the slowest predicate in the query took 0.5 seconds on LibreOffice.